File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        size = len(nums)
        for i in range(size):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j = i + 1
            k = size-1
            while(j < k):
                total = nums[i] + nums[j] + nums[k]
                if total == 0:
                    ans.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while(j < k and nums[j] == nums[j-1]):
                        j += 1
                    while(j < k and nums[k] == nums[k+1]):
                        k -= 1
                elif total > 0:
                    k -= 1
                else:
                    j += 1
        return ans
        

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;

        for(int i = 0; i < n; i++){
            if(i > 0 && nums[i] == nums[i-1]) continue;
            
            int s = i+1;
            int e = n-1;
            while(s < e){
                int curr = nums[i] + nums[s] + nums[e];
                if(curr == 0){
                    ans.push_back({nums[i], nums[s], nums[e]});
                    s++, e--;
                    while(s < e && nums[s] == nums[s-1]) s++;
                    while(s < e && nums[e] == nums[e+1]) e--;
                }
                else if(curr > 0) e--;
                else s++;
            }
        }
        return ans;
    }
};